ethcore/src/engines/clique/mod.rs

```diff
@@ -181,33 +178,38 @@ pub struct Clique {
 	pub block_state_by_hash: RwLock<LruCache<H256, CliqueBlockState>>,
 	pub proposals: RwLock<HashMap<Address, VoteType>>,
 	pub signer: RwLock<Option<Box<EngineSigner>>>,
-	pub step_service: Option<StepService>,
 }
 
 impl Clique {
 	/// Initialize Clique engine from empty state.
 	pub fn new(params: CliqueParams, machine: EthereumMachine) -> Result<Arc<Self>, Error> {
-		let mut engine = Clique {
+		/// Pause before starting to step Clique
+		const INITIAL_DELAY: Duration = Duration::from_secs(5);
+		/// Step Clique at most every 2 seconds
+		const SEALING_FREQ: Duration = Duration::from_secs(2);
+
+		let engine = Clique {
 			epoch_length: params.epoch,
 			period: params.period,
 			client: Default::default(),
 			block_state_by_hash: RwLock::new(LruCache::new(STATE_CACHE_NUM)),
 			proposals: Default::default(),
 			signer: Default::default(),
 			machine,
-			step_service: None,
 		};
-		if params.period > 0 {
-			engine.step_service = Some(StepService::new());
-			let engine = Arc::new(engine);
-			let weak_eng = Arc::downgrade(&engine);
-			if let Some(step_service) = &engine.step_service {
-				step_service.start(weak_eng);
-			}
-			Ok(engine)
-		} else {
-			Ok(Arc::new(engine))
-		}
+		let engine = Arc::new(engine);
+		let weak_eng = Arc::downgrade(&engine);
+
+		thread::Builder::new().name("StepService".into())
+			.spawn(move || {
+				thread::sleep(INITIAL_DELAY);
+				loop {
+					trace!(target: "miner", "StepService: triggering sealing");
+					weak_eng.upgrade().map(|x| x.step());
```

Comment:
Shall we print a warning and actually terminate the thread?

Reply:
Now I think about it, it might be better to actually take the time calling `step()` took into account, i.e.:
Otherwise if `step` takes a long time (due to lock contention or sth), we might get quite out of sync.
```rust
loop {
  let mut next_step_at = Instant::now() + SEALING_FREQ;
  weak_eng.upgrade().map(|x| x.step());
  let now = Instant::now();
  if now < next_step_at {
    thread::sleep(next_step_at - now);
  }
}
```

Reply:
You mean if the `upgrade()` fails? I read the code as "it probably doesn't matter if it skips a step" but I don't actually know. @niklasad1 ?

Reply:
@dvdplm if `upgrade` fails once it will never recover (cause it means that `Engine` has been dropped), so the loop will be infite, might be better to print a warning and terminate the thread.

Reply:
> if upgrade fails once it will never recover (cause it means that Engine has been dropped),

Got it. I misread the semantics of `upgrade`, thought it could recover. However, my question becomes: if the Engine is dropped, doesn't that mean that the process is exiting (so the thread is getting reaped anyway)? Put differently: is there a case where the Engine is dropped but we're not shutting down? I thought not.

Reply:
> take the time calling `step()` took into account,

Very good pt, fixing! :)

Reply:
> However, my question becomes: if the Engine is dropped, doesn't that mean that the process is exiting (so the thread is getting reaped anyway)? Put differently: is there a case where the Engine is dropped but we're not shutting down? I thought not.

Yeah, that should be the case, but IMHO it's better to handle that explicitly - that will make the code more future proof.
Imagine that in the future we move some other stuff to this thread and not terminating it will prevent other drops.